hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/RootProxyWorldUpdater.java

```diff
@@ -64,21 +72,53 @@ public BaseProxyWorldUpdater(
      *     <li>For each increase in storage size, calculates rent and tries to charge the allocating contract.</li>
      *     <li>"Rewrites" the pending storage changes to preserve per-contract linked lists.</li>
      * </ol>
+     *
+     * @throws ResourceExhaustedException if the storage size limit is exceeded or rent cannot be paid
      */
     @Override
     public void commit() {
-        // Get the pending changes and summarize their effects on size
+        // Validate the effects on size are legal
         final var changes = evmFrameState.getStorageChanges();
         final var sizeEffects = summarizeSizeEffects(changes);
-
-        // Validate the effects on size are legal
-        storageSizeValidator.assertValid(sizeEffects.finalSlotsUsed(), scope.dispatch(), sizeEffects.sizeChanges());
+        storageSizeValidator.assertValid(sizeEffects.finalSlotsUsed(), hederaOperations, sizeEffects.sizeChanges());
         // Charge rent for each increase in storage size
         chargeRentFor(sizeEffects);
         // "Rewrite" the pending storage changes to preserve per-contract linked lists
-        storageManager.rewrite(scope, changes, sizeEffects.sizeChanges(), scopedStorage());
+        storageManager.rewrite(hederaOperations, changes, sizeEffects.sizeChanges(), hederaOperations.getStore());
 
+        // We now have an apparently valid change set, and want to capture some summary
+        // information for the Hedera record
+        createdContractIds = hederaOperations.createdContractIds();
```

Comment:
These two methods on `HederaOperations` aren't actually annotated to return `@NonNull` (which comes up below).